multimodal/src/autogluon/multimodal/optimization/losses.py

```diff
@@ -272,3 +272,73 @@ def forward(self, features_a, features_b, logit_scale, rank=0, world_size=1):
 
         total_loss = (F.cross_entropy(logits_per_a, labels) + F.cross_entropy(logits_per_b, labels)) / 2
         return total_loss
+
+
+class FocalLoss(nn.Module):
+    """
+    Focal loss based on https://github.com/AdeelH/pytorch-multi-class-focal-loss/blob/master/focal_loss.py
+
+    References:
+        [1] https://arxiv.org/abs/1708.02002
+    """
+
+    def __init__(
+        self,
+        alpha: Optional[torch.Tensor] = None,
+        gamma: Optional[float] = 2.0,
+        reduction: Optional[str] = "mean",
+        eps: Optional[float] = 1e-6,
+    ):
+        """
+
+        Parameters
+        ----------
+        alpha
+            weighting factor for each class. Should be of shape (num_classes)
+        gamma
+            the focal parameter for calculating weights on easy/hard samples
+        reduction
+            the reduction to apply to the final loss output. Default: "mean". Options:
+                "mean", "sum"
+        eps
+            epsilon for numerical stability
+        """
+        super(FocalLoss, self).__init__()
+
+        self.gamma = gamma
+        self.reduction = reduction
+        self.eps = eps
+        if alpha is not None:
+            alpha = torch.tensor(alpha)
+        self.nll_loss = nn.NLLLoss(weight=alpha, reduction="none")
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor):
+        if not torch.is_tensor(input):
+            raise TypeError("input type is not a torch.Tensor. Got {}".format(type(input)))
+        if input.ndim > 2:
+            # (N, C, d1, d2, ..., dK) --> (N * d1 * ... * dK, C)
+            num_class = input.shape[1]
+            input = input.permute(0, *range(2, input.ndim), 1).reshape(-1, num_class)
+            # (N, d1, d2, ..., dK) --> (N * d1 * ... * dK,)
+            target = target.view(-1)
+
+        pt = F.softmax(input, dim=-1)
+
+        # -alpha_t * log(pt) term
+        log_p = torch.log_softmax(input, dim=-1)
+        ce = self.nll_loss(log_p, target)
+
+        # (1 - pt)^gamma term
+        all_rows = torch.arange(input.shape[0])
+        pt = pt[all_rows, target]
+        focal_term = (1 - pt) ** self.gamma
```

Comment:
Is the `exp()` operation missing? See https://github.com/AdeelH/pytorch-multi-class-focal-loss/blob/master/focal_loss.py#L82

Reply:
oh the author used a log_pt = log_softmax(input), and then computes pt = log_pt.exp(). I was directly computing pt = softmax(input), trying to avoid numerical stability issues during the exp after log_softmax if any.